Add computeWith to interleave gmem accesses and computations

naoyam opened the pull request:
This PR adds a new scheduling primitive called `computeWith`. It's different from what we previously had as `computeWith`.

The main motivation is that in persistent fusions we often have code like this:

```
float buffer[N];
// Load everything we need
for i in N:
  buffer[i] = load_from_gmem();
// Use the buffer
for i in N:
  F(buffer[i]);
...
// Use the buffer again
for i in N:
  G(buffer[i]);
```

Here, when `F` is some non-trivial computation, like the serial Welford computation, I observed 10-15 % perf gain by converting it the code as shown below:

```
float buffer[N];
// Load and use the data
for i in N:
  buffer[i] = load_from_gmem();
   F(buffer[i]);
...
// Use the buffer again
for i in N:
  G(buffer[i]);
```

Note that `N` is a compile-time constant, so all the loops are completely unrolled. There should be nothing to block nvcc to interleave the initial memory read and the first use automatically, but apparently that doesn't seem to happen.

The existing primitives in nvFuser don't allow us to express this code pattern. `computeAt` or `inlineAt` can be used to merge the first two loops in the first example of the above, but they also mean the allocation of the buffer is inlined inside the merged loop, thus the code would become invalid as there's another use of the buffer, `G(buffer[i])`. For this persistent pattern, we always need to make sure the buffer is allocated that is commonly in the scope of all the uses, but somehow we would like to move only the expression of the buffer read into the loop where it is first used.

This effectively means we would need something similar to `store_at` in Halide. Our `computeAt` is less flexible as it dictates both the allocation point and the computation point. 

The `computeWith` primitive is to address the lack of the store-at concept in a limited manner without heavily modifying the existing primitives. It does not allow specifying the allocation point of a tensor, but rather it can be used to inline the computation of an expression at a consumer without changing the allocation point. So, for example, in the above case, `buffer->computeWith(F, -1)` would move the buffer loading into the loop of the `F` loop. 

More specifically, `TensorView` has these new fields:

```
//! Direct consumer tensors that this tensor is computed with
std::vector<TensorView*> compute_with_consumers_;
//! Position where this tensor is computed with the compute-with
//! consumer tensor. It should be always be equal or greater than
//! the computeAt position
unsigned int compute_with_pos_ = 0;
```

`compute_with_consumers_` hold a list of consumers that this tensor is computed with. It's usually just one consumer tensor, but there can be multiple tensors when a consumer has siblings. `compute_with_pos_` is the position where this tensor is computed with the consumers. Obviously, this must be always greater than or equal to `compute_at_pos_`. 

Note that at this point I'm still keeping `compute_at_pos_` to mean the allocation point. We may want to rename it to `store_at_pos_`.

I also added several functions to `TensorView`. Namely:

- `computeWith(TensorView* consumer, int consumer_pos,  bool best_effort)` to do the computeWith setting with some error checking
- `hasComputedWith` to query if this tensor is computed with any consumer
- `isComputedWith(TensorView* consumer)` to query if it's computed with `consumer`
- `getMaxComputeAtPosition()` to get the maximum of the computeAt and computeWith positions
- `getMaybeComputeAtPosition(TensorView* consumer)` to get the position where this tensor is computed from the perspective of the given consumer. If this tenso is computed with the consumer, it returns the compute-with position. Otherwise, it returns the compute-at position, which effectively means the store-at position.

The actual implementation of `computeWith` is mostly borrowed from `inlineAt`. Unlike `computeAt` but similar to `inlineAt`, it does not transform itself nor the consumer but just sets the computeWith position and the consumer tensors that it's computed with. Unlike `inlineAt`, it does not need to have the constraint of allocation of persistent buffers since it does not change allocation points, which means we can skip building of the unmappable domains using `ComputeAtRootDomainMap`. 

For concrete examples of how it's used, see the new tests in `test_gpu_computed_with.cpp`. In particular, `FusionComputeWith7` reproducers the original motivating case of this transformation.

Some more notes:

- Nothing should change unless `computeWith` is used. 
- I looked over all the uses of `getComputeAtPosition()` and changed them to `getMaxComputeAtPosition()` or `getMaybeComputeAtPosition()` when I thought necessary. I'm not super confident but nothing seems broken.
- The LOOP computeAt map is modified to map all leaf IDs that are left of the compute-with position when the consumer is the one this tensor is computed with.
- One of the major limitations is that there's no error check with the computed-with consumer. For example, in the above case, we can move the memory read at the first use but not at the second use since that would break the data dependency between the read and the first use. See `FusionComputeWith4` for a concrete example. I believe this error check would effectively require almost the same analysis as the expression sorting, and in fact the test fails at the expression sorting. Ideally, invalid computeWith like this should be detected at the time when a tensor is computed with, not when its fusion is lowered, but at this point, I don't see it's a must to get some progress with the outer persistent problem.

Commit:
computeWith

Commit:
Update other uses of getComputeAtPos

Commit:
update test

Commit:
Support computeWith of multiple outputs

Commit:
computeWith and siblings

Commit:
Cleanup

Commit:
fix

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/inlining.cpp:
```diff
@@ -11,12 +11,18 @@ namespace fuser {
+    bool compute_at_only)
     : uninlinable_ids_(uninlinable_ids) {
-  buildUnmappableDims();
+  buildUnmappableDims(compute_at_only);
```
nit: Why not just
```
  // When used for computeAt only, i.e., without inlining storeAt
  // positions, the restriction below does not apply
  if (compute_at_only) {
    buildUnmappableDims();
  }
```

naoyam replied:
That should work too, but I also thought `buildUnmappableDims` could populate some other IDs that should not be mapped, so I didn't want to completely skip calling `buildUnmappableDims()`. Not something I have a strong preference.

zasdfgbnm replied:
That's a good point. It might probably be still needed if we want to fix https://github.com/csarofeen/pytorch/issues/2017, I will think more about this, but for this PR, I am OK with leaving it as is.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/inlining.h:
```diff
@@ -22,7 +22,7 @@ class MaxPosCalculator {
   // Iterate through all TVs and collect the dimensions of each TV that don't
   // map to all its consumer TVs.
-  void buildUnmappableDims();
+  void buildUnmappableDims(bool compute_at_only);
```
It would be nice to refactor `MaxPosCalculator` to pull out `unmappable_dims_`, since it is effectively the only data it has. You don't have to do it in this PR, I will write a followup for cleanup once this is merged.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/ir_interface_nodes.h:
```diff
@@ -479,6 +479,54 @@ class TORCH_CUDA_CU_API TensorView : public Val {
+  }
+
+  //! Return the tensor with which this tensor is computed. nullptr is
+  //! returned if no computeWith is set
```
I think you meant the following?
```
  //! Return the tensors with which this tensor is computed. empty vector is
  //! returned if no computeWith is set
```

naoyam replied:
Oh, thanks. The comment was originally written when `compute_with_consumers_` is just `TensorView*`. Will update the comment.

zasdfgbnm reviewed (commented):
Very clean solution for computing and allocating at different positions! I have just started, posting a few comments for now.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:
```diff
@@ -79,12 +79,17 @@ void validateDoubleBufferedTensor(const TensorView* tv) {
   auto producer = def->input(0)->as<TensorView>();
   TORCH_INTERNAL_ASSERT(
-      producer->getComputeAtPosition() <= double_buffer_pos,
+      producer->getMaybeComputeAtPosition(tv) <= double_buffer_pos,
```
🤔 Not related to this PR, but can we just compare a position in producer with a position in consumer directly? What if the producer is `T0[b,b,b,b,b,b,b,b,b,I1,I2]` and `tv` is `T1[I1,I2] = squeeze(T0)` and the computeAt position of producer is after I1?

naoyam replied:
That's a good point. This seems to need a fix.

naoyam replied:
Tracked here #2158

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/lower_sync_information.cpp:
```diff
@@ -201,15 +201,15 @@ bool useSameIndex(
-          producer_tv->domain()->domain().end(),
-          producer_id));
-  if (producer_id_pos < producer_tv->getComputeAtPosition()) {
+  // If the producer ID is mapped with any of the consumer IDs, the
+  // indexing is done with the corresponding consumer ID
+  if (std::any_of(
+          consumer_tv->domain()->domain().begin(),
+          consumer_tv->domain()->domain().end(),
+          [&](IterDomain* consumer_leaf_id) {
+            return ca_map.areMapped(
+                consumer_leaf_id, producer_id, IdMappingMode::LOOP);
+          })) {
```
Just out of curiosity, is this equivalent to the following?
```C++
  auto producer_id_pos = std::distance(
      producer_tv->domain()->domain().begin(),
      std::find(
          producer_tv->domain()->domain().begin(),
          producer_tv->domain()->domain().end(),
          producer_id));
  if (producer_id_pos < producer_tv->getMaybeComputeAtPosition(consumer_tv) {
    return true;
  }
```
I am just asking, not suggesting you to write it like above.

naoyam replied:
Yes, it should be.

naoyam commented:
As a concrete data point, in one of the outer welford tests in PR #1772, `FusionGridPersistWelfordOuterNormLike256x14x512_CUDA`, it's around 330-340 GB/s on Titan RTX:

```
kernel1 run in 0.316224 ms, achieved: 324.961 GB/s
kernel1 run in 0.303104 ms, achieved: 339.027 GB/s
kernel1 run in 0.302304 ms, achieved: 339.924 GB/s
kernel1 run in 0.305152 ms, achieved: 336.752 GB/s
kernel1 run in 0.301056 ms, achieved: 341.333 GB/s
kernel1 run in 0.303264 ms, achieved: 338.848 GB/s
kernel1 run in 0.296768 ms, achieved: 346.265 GB/s
kernel1 run in 0.300992 ms, achieved: 341.406 GB/s
kernel1 run in 0.293824 ms, achieved: 349.735 GB/s
kernel1 run in 0.30048 ms, achieved: 341.988 GB/s
kernel1 run in 0.300416 ms, achieved: 342.061 GB/s
```

With `computeWith` used as shown [here](https://github.com/csarofeen/pytorch/blob/grid_persistent_normalization/torch/csrc/jit/codegen/cuda/test/test_gpu_outer_reduction.cpp#L1615), I got around 390 GB/s:

```
kernel1 run in 0.278368 ms, achieved: 369.153 GB/s
kernel1 run in 0.263232 ms, achieved: 390.38 GB/s
kernel1 run in 0.261792 ms, achieved: 392.527 GB/s
kernel1 run in 0.262144 ms, achieved: 392 GB/s
kernel1 run in 0.262912 ms, achieved: 390.855 GB/s
kernel1 run in 0.262144 ms, achieved: 392 GB/s
kernel1 run in 0.262112 ms, achieved: 392.048 GB/s
kernel1 run in 0.261696 ms, achieved: 392.671 GB/s
kernel1 run in 0.265312 ms, achieved: 387.319 GB/s
kernel1 run in 0.258336 ms, achieved: 397.778 GB/s
kernel1 run in 0.26368 ms, achieved: 389.716 GB/s
```

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/transform_replay.cpp:
```diff
@@ -880,7 +880,7 @@ bool validateDomain(TensorView* tv, TensorDomain* new_td) {
       BestEffortReplay::findFirstMismatchedID(tv->domain(), new_td);
   return first_mismatch >= (int)tv->getMaxProducerPosition() &&
-      first_mismatch >= (int)tv->getComputeAtPosition();
+      first_mismatch >= (int)tv->getMaxComputeAtPosition();
```
I think this should be `first_mismatch >= (int)tv->getComputeAtPosition()`? During transform propagation, the new tensor domain should neither violate the compute at position nor the compute with position, and compute at position is always stricter here.

naoyam replied:
The compute-with position is always greater or equal to the compute-at position, so I think this is correct. We want to make sure any mismatch only happens left of the compute-at and compute-with positions.

zasdfgbnm replied:
Oh yeah, nvm

zasdfgbnm reviewed (commented):
posting more

Commit:
Fix comment

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/test/test_gpu_compute_with.cpp:
```diff
@@ -0,0 +1,451 @@
+#if defined(USE_CUDA)
```
Thank you for creating a new test file :)

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -340,6 +342,12 @@ void TensorView::inlineAt(
+  // computeWith position, reset the computeWith setting
+  if (compute_at_pos_ >= compute_with_pos_) {
+    resetComputeWith();
+  }
```
I think we should do this before `updateMaxProducerPosition`? Otherwise, in `getConsumerPosAlignedToProducerCA`, the `producer->getMaybeComputeAtPosition(consumer)` will get an inconsistent result.

naoyam replied:
That's a good point. Will update.

naoyam replied:
Updated

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -427,6 +436,126 @@ TensorView* TensorView::computeAt(
+      ": ",
+      consumer_pos);
+
+  // Find the corresponding position in this tensor
```
In terms of interface, which do you think makes more sense? `computeAt` uses consumer position, but `inlineAt` uses this position.

naoyam replied:
`computeAt` as well as `computeWith` have a consumer parameter, and I thought it'd be more intuitive to have the position parameter relative to the consumer.

`inlineAt`, on the other hand, doesn't reference any particular consumer, and it makes sense that the position references its domain.

That said, the difference does seem confusing. If we wanted to have a uniform position semantic, I'd change `computeWith` to take a position in its domain.

zasdfgbnm replied:
I have no idea on which will be more convenient to use in the scheduler, and I have no strong opinion about which should be used here.

naoyam replied:
Most likely, the best-effort option should just work, so I think this difference won't matter much. I'm leaning toward changing this to the position in its own tensor as that's the position we ultimately save in each tensor (i.e., `compute_with_pos_`)

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -427,6 +436,126 @@ TensorView* TensorView::computeAt(
+      "Position must be right of the computeAt position. Position: ",
+      this_pos,
+      ", computeAt position: ",
+      getComputeAtPosition());
```
I think this should be
```C++
if (this_pos < getComputeAtPosition()) {
  return;
}
```
so that it is consistent to the behavior below?
```
  // If it's already set to be computed with the consumer and the
  // position is higher, nothing to change
```

naoyam replied:
I think we shouldn't simply ignore a position that is left of the computeAt position, but that's just my feeling. I'll update the comment.

naoyam replied:
Comment updated

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -427,6 +436,126 @@ TensorView* TensorView::computeAt(
+      std::find(consumers.begin(), consumers.end(), consumer) !=
+          consumers.end(),
+      "Not a direct consumer: ",
+      consumer->toString());
```
Do we need to support the following use case?
```
T1, T2, T3 = Welford(T0)
T4 = set(T1)
T2->computeWith(T4, ...) // T4 is not a direct consumer of T2
```

naoyam replied:
I don't think so as I don't see why we can't just do `T1->computeWith(T4, ...)` instead, unless that's too restrictive.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/test/test_gpu_compute_with.cpp:
```diff
@@ -0,0 +1,451 @@
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor t0 = at::randn({dimx}, options);
+  at::Tensor cg_output = at::empty({dimx}, options);
+  at::Tensor t3_output = at::empty_like(cg_output, options);
```
Is `t3_output` used?

naoyam replied:
No

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/test/test_gpu_compute_with.cpp:
```diff
@@ -0,0 +1,451 @@
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion, {t0});
+  fe.runFusion({t0}, {cg_output});
```
Why not just
```
auto cg_outputs = fe.runFusion({t0});
```
?

naoyam replied:
Not sure. Just copied an existing old softmax test. Updated.

Commit:
Reset compute-with before updating max-prod-pos

Commit:
Mention it's an error to attemp todo computeWith at a position left of CA

Commit:
test cleanup

csarofeen commented on torch/csrc/jit/codegen/cuda/compute_at_map.cpp:
```diff
@@ -18,9 +18,13 @@ namespace {
-bool idIsAComputeAtLeafDomain(IterDomain* id, TensorView* tv) {
+bool idIsAComputeAtLeafDomain(
+    IterDomain* id,
+    TensorView* tv,
```
nit: producer_tv

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp:
```diff
@@ -333,30 +339,30 @@ class ExprSegmentationSorter {
-//    os << "}Group End\n";
-//   return os;
-// }
+std::string ExprGroup::toString() const {
```
Clang tidy used to complain when I left the debug printing uncommented, don't remember why. Are you not getting warnings while leaving this in?

naoyam replied:
Apparently not.

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_interface_nodes.h:
```diff
@@ -503,6 +552,11 @@ class TORCH_CUDA_CU_API TensorView : public Val {
       TensorView* tv,
       const std::vector<int>& axes);
 
+  void resetComputeWith() {
```
nit: `clearComputeWith`

csarofeen reviewed (commented):
My main question here is what would be the disadvantage of making compute at relative to the producer instead of a consumer. Right now it seems we're trying to make sure it gets inlined directly into a specific consumer, but is there a practical limitation of making it inlined relative to all consumers, and opportunistically picking the first consumer that lower expr sort returns during lowering?

I'm generally concerned about adding a sorting constraint (as you've identified) without having a mechanism to guarantee at scheduling time it can be sorted. If we instead let it sort the same way it would without the computeWith then just inline into the first consumer, it seems like it would achieve what you're looking for without opportunity to break sorting.

Commit:
PR feedback

naoyam commented:
> My main question here is what would be the disadvantage of making compute at relative to the producer instead of a consumer. Right now it seems we're trying to make sure it gets inlined directly into a specific consumer, but is there a practical limitation of making it inlined relative to all consumers, and opportunistically picking the first consumer that lower expr sort returns during lowering?
> 
> I'm generally concerned about adding a sorting constraint (as you've identified) without having a mechanism to guarantee at scheduling time it can be sorted. If we instead let it sort the same way it would without the computeWith then just inline into the first consumer, it seems like it would achieve what you're looking for without opportunity to break sorting.

That's interesting to think about. I'd feel reluctant to inline everything opportunistically. Maybe that's just what makes most sense, but I'd be leaning towards making it opt-in or opt-out. Maybe someday we may want to port nvFuser to CPUs, where we may want to do software pipelining, so it may not be always optimal to inline as much as possible.

I'm thinking about something like this.
- Let the user specify which tensors to inline with `computeWith` (opt-in inlining)
- When lowering, sort the expressions WITHOUT considering `computeWith` and validate if computed-with tensors can be moved without invalidating the ordering determined by the sorting.

csarofeen commented:
> I'm thinking about something like this.
> Let the user specify which tensors to inline with computeWith (opt-in inlining)
> When lowering, sort the expressions WITHOUT considering computeWith and validate if computed-with tensors can be moved without invalidating the ordering determined by the sorting.

Yeah, this sounds fine. I was just thinking the same except not needing to specify a set of consumers (so if you set computeWith it's just with any consumer that matches that position).

naoyam commented:
> > I'm thinking about something like this.
> > Let the user specify which tensors to inline with computeWith (opt-in inlining)
> > When lowering, sort the expressions WITHOUT considering computeWith and validate if computed-with tensors can be moved without invalidating the ordering determined by the sorting.
> 
> Yeah, this sounds fine. I was just thinking the same except not needing to specify a set of consumers (so if you set computeWith it's just with any consumer that matches that position).

Ah, that seems like a safer and easier approach. Thanks for the suggestion. I'll update the PR.

zasdfgbnm commented:
What is the status of this PR? I guess the only remaining two things are: https://github.com/csarofeen/pytorch/pull/2156#issuecomment-1302468460 and https://github.com/csarofeen/pytorch/pull/2156#discussion_r1012406719 ?

naoyam commented:
Yeah, I'm going to change the interface. I'll ask reviews again when ready.

Commit:
Merge branch 'devel' into prelim_store_at

Commit:
Pick the first consumer as computeWith target

Commit:
test cleanup

Commit:
Resolve computeWith first

naoyam commented:
Changed the `computeWith` interface as we discussed. Here's an example with a softmax-like pattern:

```
TEST_F(NVFuserTest, FusionComputeWith2_CUDA) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  const int vec = 4;
  const int tidx = 128;
  const int dimx = 1000;

  auto input_tv0 = makeContigTensor(1);
  fusion.addInput(input_tv0);

  auto exp_tv1 = unaryOp(UnaryOpType::Exp, input_tv0);
  auto sum_exp_tv2 = sum(exp_tv1, {-1});
  auto bcast_sum_tv3 = broadcast(sum_exp_tv2, {true});

  auto exp_tv1_copy = unaryOp(UnaryOpType::Exp, input_tv0);

  auto output_tv4 = div(exp_tv1_copy, bcast_sum_tv3);

  fusion.addOutput(output_tv4);

  auto input_tv0_cache = input_tv0->cacheAfter();

  input_tv0->split(-1, vec);
  input_tv0->split(-2, tidx);
  MaxRootDomainInfoSpanningTree tree(input_tv0);
  TransformPropagatorWithCheck tp(input_tv0);
  tree.traverse(&tp);

  auto sum_exp_rf_tv5 = sum_exp_tv2->rFactor({-1});

  inlineMost();
  input_tv0_cache->computeWith(-1);

  input_tv0_cache->axis(0)->parallelize(ParallelType::BIDx);
  input_tv0_cache->axis(1)->parallelize(ParallelType::TIDx);
  scheduler_utils::parallelizeAllLike(input_tv0_cache);

  GpuLower gpulw(&fusion);
  // Lowering should automatcially pick the first consumer of the
  // computed-with tensor as its target
  checkComputeWith(
      gpulw.kernel(), input_tv0_cache, input_tv0_cache->nDims(), {exp_tv1});

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  at::Tensor t0 = at::randn({dimx}, options);

  FusionExecutor fe;
  fe.compileFusion(&fusion, {t0});
  auto cg_outputs = fe.runFusion({t0});

  auto aten_output = at::_softmax(t0.to(at::kDouble), -1, false);

  testValidate(&fusion, cg_outputs, {t0}, {aten_output}, __LINE__, __FILE__);
}
```

It now has just:
```
input_tv0_cache->computeWith(-1);
```
Previously, it was:
```
input_tv0_cache->computeWith(-1, exp_tv1);
```

The target consumer tensor is automatically picked when the fusion is lowered. It is the first consumer tensor as appearing in the sorted expression list.

One of the major changes I needed to do is to do the expression sorting twice when computeWith is used: first to resolve computeWith targets and once more for lowering. The computeWith resolution is done at the very beginning of the lowering, before examining the fusion for validations and other analyses. Resolving computeWith, i.e., finding the right target of computeWith affects the computeAt LOOP mappings as well as the max producer positions, so anything depending on that information must be done after computeWith resolution. 

It is also possible to reuse the sorted expression list but not always as we also replace some expressions in the Fusion container at the beginning of lowering, so the sorted list needs to be also updated. It doesn't mean we would need to do the complete sorting again but we could just update the list as necessary. I didn't try that, but I'd revisit if sorting is really costly.

Remaining TODO:
- [x] Don't create the computeAt map from scratch again after computeWith resolution.

Commit:
fix

Commit:
fix

Commit:
update ca map

Commit:
cleanup

Commit:
Merge branch 'devel' into prelim_store_at

naoyam commented:
This is ready for review again @zasdfgbnm @csarofeen

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -423,6 +432,178 @@ TensorView* TensorView::computeAt(
+
+  // If it's already set to be computed with the consumer and the
+  // position is higher, nothing to change
+  if (getComputeWithPosition() >= pos) {
```
Do we need to update consumers in some way in this call to make sure that consumers don't transform in an inconsistent way after the computeWith of their producers is set?

csarofeen replied:
It seems to me `resolveComputeWith` will simply fail if we're in an inconsistent state?

naoyam replied:
Ah, I see. Since at this point no consumer is aware of the potential computeWith into it, there's nothing to block the consumers to be transformed in an inconsistent way. I need to think about this more carefully.

csarofeen replied:
I don't think this needs to be a blocker for this PR, we might just want to more explicitly validate in lowering.

csarofeen replied:
Would be nice in the future to error on transformation attempt of the consumer.

naoyam replied:
Added `maybe_max_producer_pos_` (1c83e699fe495ec158772ee1e7eaddb8cbe443bc). Don't like the name, but don't have any other idea.

Modifying consumer domains where producer may be computed at should now throw an error.

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -423,6 +432,178 @@ TensorView* TensorView::computeAt(
+  }
+}
+
+bool TensorView::resolveComputeWith(const std::vector<Expr*>& sorted_exprs) {
```
Sorted expressions is post expr sorting, not just the topologically sorted DAG, correct?

naoyam replied:
Yes. Will add a comment

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -423,6 +432,178 @@ TensorView* TensorView::computeAt(
+    }
+
+    // First use found. Set it as the computeWith target tensor
+    std::cerr << "Resolve the computeWith target as: " << expr->toString();
```
Does this need to be cleaned up?

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -423,6 +432,178 @@ TensorView* TensorView::computeAt(
+
+    // First use found. Set it as the computeWith target tensor
+    std::cerr << "Resolve the computeWith target as: " << expr->toString();
+    std::vector<TensorView*> use_out_tvs{
```
Isn't this basically just `ir_utils::getConsumerTvs`?

naoyam replied:
Not exactly. This is just taking all TV outputs of an expression.

csarofeen replied:
Right, sorry, this is a per expression loop, I was thinking it was an across all consumers loop, but you're looking for the first ordered consumer.

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -423,6 +432,178 @@ TensorView* TensorView::computeAt(
+    }
+
+    for (auto consumer_tv : compute_with_consumers_) {
+      consumer_tv->updateMaxProducerPosition();
```
Okay, this function was a little unclear in my first read, but think I got it. It seems you're not trying to enforce that all consumers can support the given `computeWith` from all its producers. Instead you only care if the first consumer in the sorted list actually supports it.

This seems more permissive than the conservative approach I was thinking of where we would simply enforce all consumers would support the `computeWith` position of all producers. Which would have made the behavior more conservative but consistent with `computeAt` behavior during scheduling.

naoyam replied:
Updating the max producer position of all consumers doesn't seem quite right to me. Except for one consumer where the producer is inlined, the other consumers still see the producer produced as if no computeWith is done. 

I don't remember what parts of the system would be affected by the change of the max producer position, but the expr sorting is one of them. Thinking about the normalization pattern, and I don't think it would be able to resolve the dependencies of the consumer and producer positions if we update the max producer position of all the consumers.

csarofeen replied:
I would assume it would have to be `maxProducerAtPosition` and `maxProducerWithPosition`.

naoyam replied:
Yes, I'm thinking about adding something like `maybe_max_produce_position_`, which should work as a guard to prevent further transformations, it shouldn't be used for the expr sorting.

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_interface_nodes.h:
```diff
@@ -473,6 +473,56 @@ class TORCH_CUDA_CU_API TensorView : public Val {
+    return compute_with_pos_;
+  }
+
+  unsigned int getMaxComputeAtPosition() const {
```
Nit: `getMaxComputePosition`

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -537,14 +718,14 @@ TensorView* TensorView::reorder(const std::unordered_map<int, int>& old2new_) {
         new_pos,
-        " are within computeAtPosition = ",
-        getComputeAtPosition());
+        " are within computeAt(With)Position = ",
```
Nit: just use `computePosition` or `maxComputePosition` to generalize `compute[At, With]Position`

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp:
```diff
@@ -794,12 +805,17 @@ ExprGroup* ExprSegmentationSorter::makeMergedNode(
-    if (producer_of_consumer_edge->isA<TensorView>()) {
-      auto tv = producer_of_consumer_edge->as<TensorView>();
-      for (const auto tv_i : c10::irange(tv->getComputeAtPosition())) {
+    auto producer_of_consumer_edge =
```
This seems like a cleanup, but also more verbose 😆

naoyam replied:
Yes, it is verbose, but the main point here is the use of `getMaybeComputeAtPosition`:

```
producer_of_consumer_edge->getMaybeComputeAtPosition(
               consumer_of_consumer_edge)
```

csarofeen commented on torch/csrc/jit/codegen/cuda/lower2device.cpp:
```diff
@@ -458,6 +460,29 @@ void GpuLower::propagateExprInfo(const Expr* old_expr, const Expr* new_expr) {
+
+  bool updated = false;
+  for (auto val : fusion->usedMathVals()) {
+    auto tv = dynamic_cast<TensorView*>(val);
```
Nit: `filterByType`

naoyam replied:
AFAIK, `filterByType` doesn't support temporary containers, which is annoying.

csarofeen commented on torch/csrc/jit/codegen/cuda/grouped_reduction.cpp:
```diff
@@ -57,6 +57,13 @@ void validateReductionGrouping(
+  // Don't know which consumer would be computed with at this
+  // point. Just make sure all the grouped reduction outputs have the
+  // same set of consumers. This is not necessarily a required
+  // condition and could be made more flexible
```
Should reduction grouping be done after scheduling then? I'm confused what you're checking here.

naoyam replied:
Right now reduction grouping is provided as a scheduling primitive (`groupReductions(std::vector<TensorView*)`) so that schedulers can opt-in to group particular reductions. This part of the code is part of the validations done before converting multiple `ReductionOp` exprs to a single `GroupedReductionOp` expr. 

Everything happens before lowering, so there can be unresolved computeWith. Specifically, what this validation checks is that if the output of one of the grouped `ReductionOp` exprs is set to be computed-with, all of the other reduction outputs must have non-conflicting compute-with settings. That means all of the outputs have the same computeWith position. At this point, they are unresolved, so we don't know which consumer of each output will be picked as the target of the computeWith, so I simply enforced all of them must have the same consumers, which is more than enough to make the computeWith setting after grouping still valid.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower2device.cpp:
```diff
@@ -347,7 +349,7 @@ void GpuLower::lower(Fusion* fusion, DataType index_type) {
   // Reorder expressions for loop-nest generation respecting computeAt
   // relationships
-  const auto exprs_sorted = reorderExprsForComputeAt();
+  auto exprs_sorted = reorderExprsForComputeAt();
```
Little strange to me that above `resolveComputeWith(fusion_);`  calls `reorderExprsForComputeAt` then it's being called again here. Can the expression ordering change between the two calls based on the result of the resolution?

naoyam replied:
It can, but may not be as is. That's what I meant to say in the above:

> It is also possible to reuse the sorted expression list but not always as we also replace some expressions in the Fusion container at the beginning of lowering, so the sorted list needs to be also updated. It doesn't mean we would need to do the complete sorting again but we could just update the list as necessary. I didn't try that, but I'd revisit if sorting is really costly.

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_interface_nodes.h:
```diff
@@ -473,6 +473,56 @@ class TORCH_CUDA_CU_API TensorView : public Val {
+  //! which also means its computeWith needs to have been resolved, the
+  //! computeWith position is returned. Otherwise, the default computeAt
+  //! position is retured.
+  unsigned int getMaybeComputeAtPosition(const TensorView* consumer) const;
```
Nit: `getComputePosition`

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_interface_nodes.h:
```diff
@@ -473,6 +473,56 @@ class TORCH_CUDA_CU_API TensorView : public Val {
+  //! computeWith position is returned. Otherwise, the default computeAt
+  //! position is retured.
+  unsigned int getMaybeComputeAtPosition(const TensorView* consumer) const;
+
```
Should we always be using the new get functions for the compute at position? Do we want to hide the `getComputeAtPosition` call now or is it still needed?

naoyam replied:
I don't think we can hide `getComputeAtPosition` entirely, but I want to rename it something like `getStoreAtPosition`. We still need that information, for example, finding allocation points.

csarofeen commented on torch/csrc/jit/codegen/cuda/compute_at_map.cpp:
```diff
@@ -1563,6 +1567,60 @@ ComputeAtMap::getAllDisjointSetConsumers(
+      "Invalid tensor: ",
+      compute_with_tv->toString());
+
+  // Can use any consumer this tensor is computed with
```
I thought compute with is resolved after compute at map is built but this reads to me like we need to resolve the compute with before building the compute at map. I'm a little confused on the logical dependencies.

csarofeen replied:
Okay, think I understood now:

1. Compute at map is built
2. Expressions are sorted
3. Compute with (the producer) is resolved based on first consumer found in expression sorting
4. Compute at map updates the loop mappings based on compute with
5. Expressions are sorted again with the compute with information
The only step that concerns me I think is step (5). It seems to me we should assert the expressions are sorted the same way as in (2) otherwise I think that would be a logical inconsistency with the approach.

Loop map being updated in (4) I believe is to generate the right loop structure not to modify the sorted expression order.

csarofeen replied:
Didn't specify, @naoyam this is my only "sticking point" before approving.

naoyam replied:
Yes, that is correct. I'm not sure how the second sort could result in inconsistency. As long as the sorting work as intended, I believe we should be fine. The second sorting could result in different ordering as some more IterDomains are loop-mapped with updated max producer positions, but as long as it completes the soring algorithm, isn't the sorted list valid?

csarofeen replied:
Yeah, I'm okay with this. The expectation is the second sorting wouldn't have a tangible difference, some things I guess could be reordered, but I would think that shouldn't actually be the case because loop map would be updated based on the  first consumer after a producer and updating the loop map should only reinforce them being next to each other.

I'm just wondering if we should assert the first and second sorting match, but maybe given the algorithm has some degrees of freedom it might just trigger coincidentally some expr "group" to reorder with another expr "group" which is just that "degree of freedom". i.e. there may be more than one unique valid ordering.

Yeah, I'll change this to approve, I'm okay with this.

naoyam replied:
Yes, there's some freedom, and updated loop mapping and max producer position might affect the decision about which ordering to use within the freedom. Otherwise, everything should be deterministic (except for bugs of course). I'm reasonably confident that the expression sorting should be able to handle both the first and second sorting consistently.

csarofeen reviewed (commented):
Reviewed everything but the tests and only have one sticking point of confusion. I'll rely on @zasdfgbnm and @naoyam to make sure there's sufficient testing at this point.

csarofeen reviewed (approved):
LGTM, didn't review the tests.

Commit:
cleanup

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/lower2device.cpp:
```diff
@@ -347,7 +349,7 @@ void GpuLower::lower(Fusion* fusion, DataType index_type) {
 
   // Reorder expressions for loop-nest generation respecting computeAt
   // relationships
-  const auto exprs_sorted = reorderExprsForComputeAt();
```
Why do we lose a `const`?

naoyam replied:
Oops, that was necessary when I was playing a different approach. Will revert the change.

Commit:
Add `maybe_max_producer_pos_` to prevent modifications of axes that may
be potentially shared with producers through computeWith

Commit:
add back const
